README.md
# weather.sh and forecast.sh

A bash script to get the weather and forecast from OpenWeatherMap and output 
to the terminal, Openbox, or HTML 

Additionally, a way to put that information onto an image (see the `weather_image` subdirectory).

![Output example](https://raw.githubusercontent.com/uriel1998/weather.sh/master/example_output.png "Example output")

## Contents
 1. [About](#1-about)
 2. [License](#2-license)
 3. [Prerequisites](#3-prerequisites)
 4. [How to use](#4-how-to-use)
 5. [weather_image](#5-weather_image)
 6. [TODO](#6-todo)

***

## 1. About

Weather report written in Bash.

`weather.sh` gets the current weather from 
[OpenWeatherMap](http://openweathermap.org/) and displays the 
results to the terminal, HTML, or for an OpenBox pipe menu. It will 
calculate (if appropriate) the "feel like" weather by calculating the
wind chill or heat index. A great deal of basis for this script comes 
from [BashWeather](https://github.com/jdotjdot/BashWeather),
[bash-weather](https://github.com/szantaii/bash-weather),
and many more that I forgot to save the URLs of.

`forecast.sh` gets the forecast from [OpenWeatherMap](http://openweathermap.org/) 
and likewise displays the results. OpenWeatherMap forecasts are at three hour 
intervals (0800, 1100, 1400, etc). It displays all forecasts for the next 
twenty four hours, then displays the closest time forecast for the day after 
that. For example, if it's 1136 (as I'm writing this) it shows all forecasts 
through the 1100 forecast the next day, then the 1100 forecast for each 
subsequent day. Options and `.rc` file are *the same* as for `weather.sh`.

If you wish to show both together, simply execute:

`weather.sh && forecast.sh` 

## 2. License

This project is licensed under the MIT license. For the full license, see `LICENSE`.

## 3. Prerequisites

 * OpenWeatherMap API key ([http://openweathermap.org/appid](http://openweathermap.org/appid)).
 * Bash shell ≥ 4.2.
 * `bc` basic calculator for floating point arithmetic. Can be found in the 
 `bc` package on major Linux distributions.
 * `curl` command-line tool for getting data using HTTP protocol. cURL can be 
 found in the `curl` package on major Linux distributions.
 * `grep` command-line tool used for parsing downloaded XML data. `grep` can 
 be found in the `grep` package on major Linux distributions.
 * `jq` command-line tool for parsing JSON data. `jq` can be found in the `jq` 
 package on major Linux distributions.
 * `tr` command-line tool for parsing JSON data. `tr` can be found in the `tr` 
 package on major Linux distributions.
 * `awk` command-line tool for parsing JSON data. `awk` can be found in the 
 `awk` package on major Linux distributions. 

Optional: For colors in terminal, save `bashcolors` in this repository to 
`.bashcolors` in your `$HOME` directory.

## 4. How to use

Run `weather.sh` or `forecast.sh` with the appropriate commandline switches 
(below). If  the current conditions do not qualify for the heat index or wind 
chill, it is not displayed.

### weather_sh.rc

Copy (and edit, as appropriate) the `weather_sh.rc` file to `$HOME\.config\weather_sh.rc`.   
* The first line is the OpenWeatherMap API key  
* The second line is your default location. (See note below)  
* The third line is your default degree character (either `c` or `f`)  
* The fourth line is True or False depending on whether or not you want icons displayed for the weather.
* The fifth line is whether to use `bashcolors`.

### Command-line options

`weather.sh` and `forecast.sh` can be started with the following command line 
options:

 * `-k` Specifies OpenWeatherMap API key from the command-line.
 * `-l city_name` Sets the city for manual weather lookup. (see note below)
 * `-t` Output to the terminal/stdout (default if no output is specified)
 * `-h` Output HTML formatted text
 * `-o` Output OpenBox output
 * `-y` Output Conky format (no icons)
 * `-n` Terminal output without icons
 * `-p` Specify cache/temp path *with* trailing slash
 * `-f` Use imperial (farenheit, inches Hg, mph) units; default is metric
 * `-c` Use colored output in the terminal if `.bashcolors` is in the home 
 dir. Note that if you want to alter the colors, you will have to manually
 alter the script.
 
_Note: If the OpenWeatherMap API key is specified from the command-line, it 
will override the API key set in the file._

_Note: It is **STRONGLY** recommended to use the City ID from OpenWeatherMap 
instead of a city name. Instructions on finding your city's City ID 
[here](https://www.dmopress.com/openweathermap-howto/) ._


### Calling from Conky

I have a single line config for my secondary screen with the weather config 
in it:

`Now: ${execi 300 weather.sh -y} Forecast: ${execi 300 forecast.sh -y}`

The conky output is currently limited via code to just the next five outputs.

## 5. weather_image

Please see the [weather_image](https://github.com/uriel1998/weather.sh/tree/master/weather_image) subdirectory.

## 6. Todo

 * Add in sunrise/sunset
 * HTML colored output
 * Current location instead of hardcoded 
    - this is problematic due to the way the API looks up city names.


CHANGELOG.md
* [2021-10-28 18:15:38 CDT] - desc of readme tag: 1.1.0
* [2021-10-28 18:13:08 CDT] - up readmes
* [2021-10-28 18:08:05 CDT] - readme, licenses, credits for icons
* [2021-10-28 17:50:51 CDT] - readme, fixed font issue in both
* [2021-10-28 17:15:28 CDT] - with font finding information and prereq
* [2021-10-28 17:08:07 CDT] - images
* [2021-10-28 17:04:30 CDT] - choosing working
* [2021-10-28 16:30:24 CDT] - readme, realized I need to do a font sanity check...
* [2021-10-28 15:57:37 CDT] - scriptname to avoid confusion
* [2021-10-28 15:56:33 CDT] - text added
* [2021-10-28 15:48:25 CDT] - Now for help text - and a switch for that....
* [2021-10-28 15:45:57 CDT] - working, now to quiet the TTY output...
* [2021-10-28 15:32:58 CDT] - of random pull
* [2021-10-28 15:28:24 CDT] - with multiple output files working fixed icon placement for smaller width files
* [2021-10-28 14:48:39 CDT] - in shuf for sort/head combo and fixed error
* [2021-10-28 14:41:56 CDT] - in option of using find
* [2021-10-26 16:27:03 CDT] - weather_image in main branch
* [2021-10-26 16:21:29 CDT] - branch 'master' weather_image
* [2021-10-26 16:20:47 CDT] - output provided
* [2021-10-26 16:20:30 CDT] - works!
* [2021-10-26 12:00:36 CDT] - it with all the quotes
* [2021-10-26 11:44:12 CDT] - had that backwards...
* [2021-10-26 11:43:06 CDT] - smokes, I think I have it.
* [2021-10-26 11:07:02 CDT] - working weather image, now for source selection
* [2021-10-25 21:32:08 CDT] - logic sketch
* [2021-10-25 21:28:36 CDT] - ing in files for weather_image start
* [2020-12-03 07:53:02 CDT] - typo in version
* [2020-12-03 07:51:55 CDT] - changelog tag: v1.0.3
* [2020-12-03 07:50:02 CDT] - for conky output so "cache age" doesn't appear tag: v.1.0.3
* [2020-11-29 14:34:36 CDT] - 214956c N 2020-11-29 14:34:36 -0500 Steven Saus         changelog
* [2020-11-29 14:33:15 CDT] - resiliance and checks for proper city response tag: v1.0.2
* [2020-11-29 11:14:14 CDT] - changelog
* [2020-11-29 11:13:16 CDT] - remote-tracking branch 'refs/remotes/github/master' tag: v1.0.1
* [2020-10-11 22:42:57 CDT] - theme jekyll-theme-midnight
* [2020-11-29 11:10:40 CDT] - changelog
* [2020-11-29 11:09:36 CDT] - -n and -p switches and docs for both
* [2020-11-29 11:10:40 CDT] - changelog
* [2020-11-29 11:09:36 CDT] - -n and -p switches and docs for both tag: v1.0.0
* [2020-10-11 22:42:57 CDT] - theme jekyll-theme-midnight
* [2020-10-11 22:23:10 CDT] - Pages
* [2020-10-11 21:28:14 CDT] - 7759fc9 N 2020-10-11 21:28:14 -0400 Steven Saus         pages
* [2020-08-29 21:23:18 CDT] - up README;silence CURL tag: 0.9.9
* [2019-06-20 19:08:50 CDT] - conky code for both forecast and output
* [2019-05-31 11:50:15 CDT] - pesky "clear" command
* [2019-05-31 11:29:48 CDT] - variable error with HTML/Openbox forecast output
* [2019-05-26 13:02:30 CDT] - to get an inline image on three different...i mean, really.
* [2019-05-26 12:55:29 CDT] - inline image
* [2019-05-26 12:53:29 CDT] - rid of stub
* [2019-05-26 12:52:47 CDT] - image
* [2019-05-26 12:46:30 CDT] - by city ID OR name, example output as part of README
* [2019-05-26 11:40:16 CDT] - forecast.sh with colors, wrote forecast.sh into readme
* [2019-05-26 09:48:58 CDT] - colors to forecast terminal output
* [2019-05-01 09:30:25 CDT] - additional (openbox, html) outputs for forecast
* [2019-05-01 09:13:48 CDT] - forecast script, removing extra newlines on weather output
* [2019-03-24 18:01:25 CDT] - readme
* [2019-03-24 18:00:21 CDT] - name in readme
* [2019-03-24 17:58:07 CDT] - switch and rc for icon display
* [2019-03-24 17:51:43 CDT] - rc calling and loop
* [2019-03-24 17:44:51 CDT] - initialization of variables
* [2019-03-24 17:44:09 CDT] - RC
* [2019-03-24 17:33:57 CDT] - commit
* [2019-03-24 17:12:00 CDT] - commit


weather_image/weather_image.sh
#!/bin/bash

TempDir=$(mktemp -d)
SCRIPT_DIR="$(dirname "$(readlink -f "$0")")"
BlurVar=""
FD_FIND=$(which fdfind)
# Going to do programmer style counting, starting at 0 here.
NumOutput=1 
OccasionalRandom=""
RandoRun=""
FontFamily=""
ExplicitConfigFilePath=""

get_image () {
    if [ -d "$ImageDir" ];then
        if [ -f "$FD_FIND" ];then
            ImageFile=$(fdfind . "$ImageDir" --follow --type file --extension png --extension jpg | shuf -n 1 )
        fi
        if [ ! -f "${ImageFile}" ];then
            ImageFile=$(find "$ImageDir" -type f  -iname "*.jpg" -or -iname "*.png" -printf '%h\n' | shuf -n 1 )
        fi
    fi
    if [ ! -f "${ImageFile}" ];then
        ImageFile=""
    fi
}


print_help (){

    echo " # USAGE: weather_images.sh [options]"
    echo " Defaults are to source the image from pixabay and output to "
    echo " out.jpg in the script's directory with Interstate font."
    echo " ## TOGGLES:"
    echo " -b : add blur"
    echo " -r : occasionally mix in random image from internet (every third)"
    echo " -help | -? : show help text"
    echo " ## SWITCHES THAT NEED MORE INPUT AFTER THEM " 
    echo " -n ### : the number of output images to make (autonumbered)"
    echo " -d [directory] : image directory to choose from"
    echo " -i [file] : specific image file to use"
    echo " -h ### : height if sourced from pixabay"
    echo " -w ### : height if sourced from pixabay "
    echo " -o [full path] : specify output file "
    echo " -f [font family] : specify font family to use "
    echo " -c [full path] : specify config file "
    exit 0
    
}

check_fonts (){

    fontexists=""
    if [ -n "$FontFamily" ];then 
        case "$FontFamily" in  
        *\ * )
            echo " "
            echo "There is a space in your font family. Because of the way imagemagick"
            echo "and this script handle fonts, please use the *filename* of the font, by typing "
            echo "fc-list | grep -i \"$FontFamily\" and choosing the filename. "
            echo "We are now going to try to choose a random font from that family."
            echo " "
            FontFamily=$(fc-list | grep -i "$FontFamily" | grep -E ".ttf|.otf" | shuf -n 1 | awk -F ':' '{print $1}')
            fontexists=$(fc-list | grep -ci "$FontFamily")
            ;;
        *)
            fontexists=$(fc-list | grep -ci "$FontFamily")
            ;;
        esac
    fi
    if [ -z "$fontexists" ];then
        FontFamily="Interstate"
        fontexists=$(fc-list | grep -ci "$FontFamily")
        if [ -z "$fontexists" ];then
            FontFamily="Ubuntu"
            fontexists=$(fc-list | grep -ci "$FontFamily")
            if [ -z "$fontexists" ];then
                FontFamily="Arial"
                fontexists=$(fc-list | grep -ci "$FontFamily")
                
                # I've tried what was given with no dice, so I'm picking one at random
                if [ -z "$fontexists" ];then
                    FontFamily=$(fc-list | shuf -n 1 | awk -F ':' '{ print $2 }' | awk '{$1=$1;print}')
                fi
            fi
        fi
    fi
    echo "Using font $FontFamily."
}

while [ $# -gt 0 ]; do
option="$1"
    case $option
    in
        -c) 
            ExplicitConfigFilePath="$2"
            if [ ! -f "${ExplicitConfigFilePath}" ];then
                ExplicitConfigFilePath=""
            fi
            shift
            shift ;;
        -f)           
            FontFamily="$2"
            shift
            shift;;
        -r) 
            OccasionalRandom=true
            shift ;;
        -b) 
            BlurVar="True"
            shift ;;
        -n) 
            NumOutput="$2"
            shift
            shift;;
        -d) 
            ImageDir="$2"
            if [ ! -d "${ImageDir}" ];then
                ImageFile=""
            else
                get_image
                if [ ! -f "${ImageFile}" ];then
                    ImageFile=""
                fi
            fi
            shift
            shift ;;
        -i) 
            ImageFile="$2"
            if [ ! -f "${ImageFile}" ];then
                ImageFile=""
            fi
            shift
            shift ;;
        -w) 
            UnsplashWidth="$2"
            shift
            shift ;;  
        "-?"|"-help") 
            print_help
            ;;
        -h) 
            UnsplashHeight="$2"
            if [ -z "$UnsplashHeight" ];then
                print_help
            fi
            shift
            shift ;;
        -o)
            OutputFile="$2"
            shift
            shift ;;
    esac
done

if [ -z "${UnsplashHeight}" ];then
    UnsplashHeight="1080"
fi
if [ -z "${UnsplashWidth}" ];then
    UnsplashWidth="1920"
fi

check_fonts

################################################################################
# Wherein things get told to happen
################################################################################
main() {
    
    LOOP=$(printf "%03g" "$1")
    ForcePull="$2"
    if [ -z "$ImageFile" ];then 
        if [ -n "$ForcePull" ];then
            ImageFile=""
        else
            get_image
        fi
    fi
	if [ -z "${ImageFile}" ];then
        wget_bin=$(which wget)
        # Obtain source image
        execstring="${wget_bin} https://picsum.photos/${UnsplashWidth}/${UnsplashHeight}/?random -qO ${TempDir}/unsplash.jpg"
        eval "$execstring"
	else
        cp "${ImageFile}" "${TempDir}"/unsplash.jpg
    fi
    
    if [ -n "$BlurVar" ];then
        # Blur, if desired. 
        convert "${TempDir}"/unsplash.jpg -blur 0x4 "${TempDir}"/unsplash_blur.jpg
    else
        cp "${TempDir}"/unsplash.jpg "${TempDir}"/unsplash_blur.jpg
    fi
    
    ImageSize=$(identify "${TempDir}"/unsplash_blur.jpg  | awk '{print $3}')
    ImageWidth=$(echo "${ImageSize}" | awk -F 'x' '{print $1}')
    ImageHeight=$(echo "${ImageSize}" | awk -F 'x' '{print $2}')
    # because otherwise the text gets squashed
    if [ "$ImageWidth" -le 1366 ];then
        TextWidth=$(( "$ImageWidth" / 3 ))
        TextWidth=$(( "$TextWidth" * 2 ))
    else
        TextWidth=$(( "$ImageWidth" / 2 ))
    fi
    TextHeight=$(( "$ImageHeight" / 2 ))
    
    
    
    # Get our text and make it into an image
    if [ -n "$ExplicitConfigFilePath" ];then
        DataInfo=$("${SCRIPT_DIR}"/weather_image_helper.sh -r "$ExplicitConfigFilePath" | grep -v "Cache")
    else
        DataInfo=$("${SCRIPT_DIR}"/weather_image_helper.sh | grep -v "Cache")
    fi
    # Sometimes it feeds us two icons worth of data; take the first.
    IconData=$(echo "$DataInfo" | awk '{print $1}' | head -1)
    TextData=$(echo "$DataInfo" | tail -6)
    cp "${SCRIPT_DIR}"/icons/"$IconData".png "${TempDir}"/WeatherIcon.png
	/usr/bin/convert -background none -fill white -stroke black -strokewidth 2 -gravity Southeast -font "${FontFamily}" -size "$TextWidth"x"$TextHeight" \
          caption:"$TextData" \
          -gravity Southwest \
          "${TempDir}"/TextImage.png

    # Applying the appropriate icon to the image.  This has to be done in 
    # steps for the transparency to keep working.
    /usr/bin/composite -gravity Center "${TempDir}"/WeatherIcon.png -gravity Southwest "${TempDir}"/TextImage.png "${TempDir}"/Text_Icon.png
   
   # Applying the text and icon to the base image.
    /usr/bin/composite -gravity Southeast "${TempDir}"/Text_Icon.png "${TempDir}"/unsplash_blur.jpg "${TempDir}"/weather.jpg
    if [ -z "${OutputFile}" ]; then
        cp "${TempDir}"/weather.jpg "${SCRIPT_DIR}"/output_"${LOOP}".jpg
    else
        if [ "$NumOutput" = 1 ];then
            cp "${TempDir}"/weather.jpg "${OutputFile}"
        else
            OutExt="${OutputFile##*.}"
            Out_File="${OutputFile%.*}"
            cp "${TempDir}"/weather.jpg "${Out_File}"_"${LOOP}"."${OutExt}"
        fi
    fi
    ImageFile=""

}

################################### CONTROL PORTION ##########################

    n=0
    NumOutput=$(( "$NumOutput" ))
    while [ $n -lt "$NumOutput" ]; do
        if [ "$OccasionalRandom" = "true" ] && [ $n -ne 0 ];then
            RandoRun=""
            if ! (( "$n" % 3 )) ; then
                # forcing random pull
                RandoRun="true"
                ImageFile=""
            fi
        fi
        printf "%s" "#"
        main $n $RandoRun
        n=$(( n+1 ))
    done
    printf "\n"

rm "${TempDir}"/TextImage.png
rm "${TempDir}"/unsplash_blur.jpg
rm "${TempDir}"/unsplash.jpg
rm "${TempDir}"/WeatherIcon.png
rm -rf "${TempDir}"


weather.sh
#!/bin/bash

# TODO: Add in sunrise/sunset calcuations

apiKey=""
defaultLocation=""
Conky="False"
OpenBox="False"
Terminal="False"
HTML="False"
degreeCharacter="c"
data=0
lastUpdateTime=0
FeelsLike=0
dynamicUpdates=0
UseIcons="True"
colors="False"
CityID="True"

ConfigFile="$HOME/.config/weather_sh.rc"

if [ "$1" == "-r" ];then
    shift
    ConfigFile="$1"
    shift
fi

if [ -f "$ConfigFile" ];then
    readarray -t line < "$ConfigFile"
    apiKey=${line[0]}
    defaultLocation=${line[1]}
    degreeCharacter=${line[2]}
    UseIcons=${line[3]}
    temp=${line[4]}
    if [ "$temp" = "True" ];then
        if [ -f "$HOME/.bashcolors" ];then
            source "$HOME/.bashcolors"
            colors="True"
        else
            colors=""
        fi
    else
        colors=""
    fi
fi

while [ $# -gt 0 ]; do
option="$1"
    case $option
    in
    -k) apiKey="$2"
    shift
    shift ;;
    -l) defaultLocation="$2"
    shift
    shift ;;
    -d) dynamicUpdates=1
    shift ;;
    -t) Terminal="True"
    shift ;;
    -h) HTML="True"
    shift ;;
    -o) OpenBox="True"
    shift ;;    
    -y) Conky="True"
    shift ;;    
    -f) degreeCharacter="f"
    shift ;;
    -p) CachePath="$2"
    shift
    shift ;;
    -n) UseIcons="False"
    shift ;;
    -c) 
        if [ -f "$HOME/.bashcolors" ];then
            source "$HOME/.bashcolors"
            colors="True"
        fi
    shift ;;
    esac
done

if [ -z "${CachePath}" ];then 
    dataPath="/tmp/wth-$defaultLocation.json"
else
    dataPath="${CachePath}/wth-$defaultLocation.json"
fi

if [ -z $apiKey ];then
    echo "No API Key specified in rc, script, or command line."
    exit
fi

#Is it City ID or a string?
case $defaultLocation in
    ''|*[!0-9]*) CityID="False" ;;
    *) CityID="True" ;;
esac

if [ ! -e $dataPath ];
then
    touch $dataPath
    #The API call is different if city ID is used instead of string lookup
    if [ "$CityID" = "True" ];then
        data=$(curl "http://api.openweathermap.org/data/2.5/weather?id=$defaultLocation&units=metric&appid=$apiKey" -s )
    else
        data=$(curl "http://api.openweathermap.org/data/2.5/weather?q=$defaultLocation&units=metric&appid=$apiKey" -s )
    fi
    echo $data > $dataPath
else
    data=$(cat $dataPath)
fi
lastUpdateTime=$(($(date +%s) -600))

while true; do
    lastfileupdate=$(date -r $dataPath +%s)
    if [ $(($(date +%s)-$lastfileupdate)) -ge 600 ];then
        if [ "$CityID" = "True" ];then
            data=$(curl "http://api.openweathermap.org/data/2.5/weather?id=$defaultLocation&units=metric&appid=$apiKey" -s )
        else
            data=$(curl "http://api.openweathermap.org/data/2.5/weather?q=$defaultLocation&units=metric&appid=$apiKey" -s )
        fi
        echo $data > $dataPath
    else
        if [ "$Conky" != "True" ];then 
            echo "Cache age: $(($(date +%s)-$lastfileupdate)) seconds."
        fi
    fi
    check=$(echo "$data" | grep -c -e '"cod":"40')
    check2=$(echo "$data" | grep -c -e '"cod":"30')
    sum=$(( $check + $check2 ))
    if [ $sum -gt 0 ];then
        exit 99
    fi
    if [ $(($(date +%s)-$lastUpdateTime)) -ge 600 ]; then
        lastUpdateTime=$(date +%s)
        Station=$(echo $data | jq -r .name)
        Lat=$(echo $data | jq -r .coord.lat)
        Long=$(echo $data | jq -r .coord.lon)
        Country=$(echo $data | jq -r .sys.country)

        ####################################################################
        # Current conditions (and icon)
        ####################################################################
        if [ "$UseIcons" = "True" ];then
            icons=$(echo $data | jq -r .weather[].icon | tr '\n' ' ')
            iconval=${icons%?}
            case $iconval in
                01*) icon="☀️";;
                02*) icon="🌤";;
                03*) icon="🌥";;
                04*) icon="☁";;
                09*) icon="🌧";;
                10*) icon="🌦";;
                11*) icon="🌩";;
                13*) icon="🌨";;
                50*) icon="🌫";;
            esac
        else
            icon=""
        fi
        ShortWeather=$(echo $data | jq -r .weather[].main | tr '\n' ' '| awk '{$1=$1};1' )
        LongWeather=$(echo $data | jq -r .weather[].description | sed -E 's/\S+/\u&/g' | tr '\n' ' '| awk '{$1=$1};1' )

        ####################################################################
        # Temperature
        ####################################################################
        tempinc=$(echo $data | jq -r .main.temp| awk '{$1=$1};1' )   
        RawTemp=$(echo $data | jq -r .main.temp| awk '{$1=$1};1' )
        temperature=$tempinc
        if  [ "$degreeCharacter" = "f" ]; then
            temperature=$(echo "scale=2; 32+1.8*$tempinc" | bc| awk '{$1=$1};1' )
        fi
        
        ####################################################################
        # Parse Wind Info
        ####################################################################
        wind=$(echo $data | jq .wind.deg)
        winddir=$((2193-(${wind%.*}+45)/90))
        if [ $winddir -eq 2192 ]; then
            winddir=2190
        elif [ $winddir -eq 2190 ];then
            winddir=2192
        else
            :
        fi
        RawWindSpeed=$(echo $data | jq .wind.speed)
        WindSpeed=$(echo $data | jq .wind.speed)
        WindGusts=$(echo $data | jq .wind.gust)
        
        #Conversion
        if  [ "$degreeCharacter" = "f" ]; then
            WindSpeed=$(echo "scale=2; $WindSpeed*0.6213712" | bc | xargs printf "%.2f"| awk '{$1=$1};1' )
            WindGusts=$(echo "scale=2; $WindGusts*0.6213712" | bc | xargs printf "%.2f"| awk '{$1=$1};1' )
            windunit="mph"
        else
            WindGusts=$(echo "scale=2; $WindGusts*1" | bc| awk '{$1=$1};1' )
            windunit="kph"
        fi        

        Humidity=$(echo $data | jq .main.humidity| awk '{$1=$1};1' )
        CloudCover=$(echo $data | jq .clouds.all| awk '{$1=$1};1' )

        ####################################################################
        # Feels Like Calculations
        # Using the raw metric value for criteria, then converting later
        ####################################################################
        # Wind Chill
        ####################################################################
        if (( $(bc -l<<<"$RawWindSpeed > 4.5") )); then #windspeed criteria for windchill
            if (( $(bc -l<<<"$RawTemp< 11") )); then #temp criteria for windchill
                FeelsLike=1
                if [ "degreeCharacter" = "f" ];then
                    WindSpeedExp=$(echo "e(0.16*l($WindSpeed))" | bc -l )
                    FeelsLikeTemp=$(echo "scale=2; 35.74 + 0.6215*$temperature - 35.75*$WindSpeedExp + 0.4275*$temperature*$WindSpeedExp" | bc | xargs printf "%.2f"| awk '{$1=$1};1' )
                else
                    WindSpeedExp=$(echo "e(0.16*l($WindSpeed))" | bc -l )
                    FeelsLikeTemp=$(echo "scale=2; 13.12 + 0.6215*$temperature - 11.37*$WindSpeedExp + 0.3965*$temperature*$WindSpeedExp" | bc | xargs printf "%.2f"| awk '{$1=$1};1' )
                fi
            fi
        fi

        ####################################################################
        # Heat Index
        # I can only find Farenheit calcuations, so....
        ####################################################################
        if  [ "$degreeCharacter" = "c" ]; then
            HITemp=$(echo "scale=2; 32+1.8*$tempinc" | bc)
        else
            HITemp=$RawTemp
        fi
        if (( $(bc -l<<<"$HITemp> 79") )); then #temp criteria for heat index
            FeelsLike=1
            FeelsLikeTemp=$(echo "scale=2;0.5 * ($HITemp + 61.0 + (($HITemp-68.0)*1.2) + ($Humidity*0.094))" | bc| awk '{$1=$1};1' )
            if [ "$degreeCharacter" = "c" ];then
                FeelsLikeTemp=$(echo "scale=2; ($FeelsLikeTemp-32) / 1.8" | bc| awk '{$1=$1};1' )

            fi
        fi

        ####################################################################
        # Pressure Data
        ####################################################################
        pressure=$(echo $data | jq .main.pressure)
        if  [ "$degreeCharacter" = "f" ]; then
            pressure=$(echo "scale=2; $pressure/33.863886666667" | bc | awk '{$1=$1};1' )
            pressureunit="in"
        else
            pressureunit="hPa"
        fi
    fi
    AsOf=$(date +"%Y-%m-%d %R" -d @$lastfileupdate) 
    if [ "$OpenBox" = "False" ];then
        if [ "$HTML" = "False" ];then
            if [ "$Conky" = "False" ];then
                Terminal="True"
            fi
        fi
    fi
    if [ "$Terminal" = "True" ];then
        if [ "$colors" = "True" ]; then
            echo "Station: $Station, $Country $Lat / $Long"
            echo "As Of: ${YELLOW}$AsOf ${RESTORE}"  
            echo "Right Now: ${CYAN}$icon $LongWeather${RESTORE}"
            #echo "$icon $ShortWeather"
            echo "Temp: ${CYAN}$temperature°${degreeCharacter^^}${RESTORE}"
            if [ "$FeelsLike" = "1" ];then
                echo "Feels Like: ${RED}$FeelsLikeTemp°${degreeCharacter^^}${RESTORE}"
            fi
            echo "Pressure: ${GREEN}$pressure$pressureunit${RESTORE}"
            if [ "$UseIcons" = "True" ];then
                echo -e \\u$winddir "${MAGENTA}$WindSpeed$windunit${RESTORE} Gusts: ${MAGENTA}$WindGusts$windunit${RESTORE}"
            else
                echo "Wind: ${MAGENTA}$WindSpeed$windunit${RESTORE} Gusts: ${MAGENTA}$WindGusts$windunit${RESTORE}"
            fi
            echo "Humidity: ${GREEN}$Humidity%${RESTORE}"
            echo "Cloud Cover: ${GREEN}$CloudCover%${RESTORE}"        
        else
            echo "Station: $Station, $Country $Lat / $Long"
            echo "As Of: $AsOf "  
            echo "Right Now: $icon $LongWeather"
            #echo "$icon $ShortWeather"
            echo "Temp: $temperature°${degreeCharacter^^}"
            if [ "$FeelsLike" = "1" ];then
                echo "Feels Like: $FeelsLikeTemp°${degreeCharacter^^}"
            fi
            echo "Pressure: $pressure$pressureunit"
            echo -e \\u$winddir "$WindSpeed$windunit Gusts: $WindGusts$windunit"
            echo "Humidity: $Humidity%"
            echo "Cloud Cover: $CloudCover%"
        fi
    fi
    if [ "$Conky" = "True" ]; then
        if [ "$colors" = "True" ]; then
            bob=$(echo "$ShortWeather $temperature°${degreeCharacter^^}")
            if [ "$FeelsLike" = "1" ];then
                bob=$(echo "$bob/$FeelsLikeTemp°${degreeCharacter^^}")
            fi
        else
            bob=$(echo "$ShortWeather $temperature°${degreeCharacter^^}")
            if [ "$FeelsLike" = "1" ];then
                bob=$(echo "$bob/$FeelsLikeTemp°${degreeCharacter^^}")
            fi

        fi
        echo "$bob"
    fi
    if [ "$OpenBox" = "True" ]; then
        echo '<openbox_pipe_menu>' 
        echo '<separator label="Weather" />' 
        printf '<item label="Station: %s, %s" />\n' "$Station" "$Country"  
        printf '<item label="As of %s" />\n' "$AsOf" 
        printf '<item label="Now: %s %s" />\n' "$icon" "$LongWeather" 
        printf '<item label="Temp: %s%s" />\n' "$temperature" "°${degreeCharacter^^}" 
        if [ "$FeelsLike" = "1" ];then
            printf '<item label="Feels Like: %s%s" />\n' "$FeelsLikeTemp" "°${degreeCharacter^^}" 
        fi
        printf '<item label="Pressure: %s%s" />\n' "$pressure" "$pressureunit" 
        printf '<item label="Wind: %s%s Gusts: %s%s" />\n'  "$WindSpeed" "$windunit" "$WindGusts" "$windunit" 
        printf '<item label="Humidity: %s%%" />\n' "$Humidity" 
        printf '<item label="Cloud Cover: %s%%" />\n' "$CloudCover" 
        echo '</openbox_pipe_menu>' 
    fi
    if [ "$HTML" = "True" ];then
        echo "Station: $Station, $Country $Lat / $Long <br  />"  
        echo "As Of: $AsOf <br  />"  
        echo "Current Conditions: $icon $LongWeather <br  />" 
        #echo "$icon $ShortWeather" 
        echo "Temp: $temperature °${degreeCharacter^^} <br  />" 
        if [ "$FeelsLike" = "1" ];then
            echo "Feels Like: $FeelsLikeTemp °${degreeCharacter^^} <br  />"
        fi
        echo "Pressure: $pressure $pressureunit <br  />" 
        echo -e \\u$winddir "$WindSpeed$windunit Gusts: $WindGusts$windunit <br  />" 
        echo "Humidity: $Humidity% <br  />" 
        echo "Cloud Cover: $CloudCover% <br  />"     
    fi
    if [ $dynamicUpdates -eq 0 ];then
        break
    fi    
done
